graph/patterns/upsert.md

```diff
@@ -0,0 +1,173 @@
+* The client-provided alternate key must be immutable after being set. If its value is null then it should be settable as a way to backfill existing resources for use in IaC scenarios.
+* API producers could use `PUT` operations to create or update, but generally this approach is not recommended due to the destructive nature of `PUT`'s replace semantics.
+* API producers could to use `UPSERT` with a primary (client-provided) key and this may be appropriate for some scenarios. However, the recommendation is for resources to support creation using `POST` and a service-generated primary key, for consistency reasons.
+* API producers may annotate entity sets, singletons and collections to indicate that entities can be "upserted". The example below shows this annotation for the `groups` entity set.  
+
+```xml
+<EntitySet Name="groups" EntityType="microsoft.graph.group">
+  <Annotation Term="Org.OData.Capabilities.V1.UpdateRestrictions">
+    <Record>
+      <PropertyValue Property="Upsertable" Bool="true"/>
+    </Record>
+  </Annotation>
+</EntitySet>
+```
+
+## Examples
+
+For these examples we'll use the `group` entity type, which defines both a primary (service-generated) key (`id`) and an alternate (client-provided) key (`uniqueName`).  
+
+```xml
+<EntityType Name="group">
+  <Key>
+    <PropertyRef Name="id"/>
+  </Key>
+  <Property Name="id" Type="Edm.String"/> 
+  <Property Name="uniqueName" Type="Edm.String"/>
+  <Property Name="displayName" Type="Edm.String"/>
+  <Property Name="description" Type="Edm.String"/> 
+  <Annotation Term="OData.Community.Keys.V1.AlternateKeys">
+    <Collection>
+      <Record Type="OData.Community.Keys.V1.AlternateKey">
+        <PropertyValue Property="Key">
+        <Collection>
+            <Record Type="OData.Community.Keys.V1.PropertyRef">
+            <PropertyValue Property="Name" PropertyPath="uniqueName" />
+            </Record>
+        </Collection>
+        </PropertyValue>
+      </Record>
+    </Collection>
+  </Annotation>
+  </Property>
+</EntityType>
+```
+
+### Upserting a record (creation path)
+
+Create a new group, with a `uniqueName` of "Group157". In this case, this group does not exist.
+
+```http
+PATCH /groups(uniqueName='Group157')
+```
+
+```json
+{
+    "displayName": "My favorite group",
+    "description": "All my favorite people in the world"
+}
+```
+
+Response:
+
+```http
+201 created
+```
+
+```json
+{
+    "id": "1a89ade6-9f59-4fea-a139-23f84e3aef66",
+    "displayName": "My favorite group",
+    "description": "All my favorite people in the world",
+    "uniqueName": "Group157"
+}
+```
+
+### Upserting a record (update path)
+
+Create a new group, with a `uniqueName` of "Group157", exactly like before. Except in this case, this group already exists. This is a common scenario in IaC, when a deployment template is re-run multiple times.
+
+```http
+PATCH /groups(uniqueName='Group157')
+```
+
+```json
+{
+    "displayName": "My favorite group",
```

Comment:
May be it make sense to provide different property values to demonstrate that PATCH will work as a simple PATCH for an existing resource.

Reply:
@OlgaPodo Did you want me to add a new scenario where I show a regular update, that changes properties?
This particular one is explicitly the same to demonstrate that the outcome is predictable, when the template is run multiple times.